### PR TITLE
Use existing cookiecutters

### LINKED ISSUE #385: .cookiecutterrc and cookiecutters_dir not working as I expected
Hi,
Here's the setup:
+ I have a ~/.cookiecutterrc as talked about here: http://cookiecutter.readthedocs.org/en/latest/advanced_usage.html
+ I also have a cookiecutters_dir at ~/.cookiecutters/ with the subdirectory cookiecutter-pypackage/.

When I try to run "cookiecutter cookiecutter-pypackage/" in ~/Projects/, I get the following error

    Traceback (most recent call last):
      File "/opt/anaconda/bin/cookiecutter", line 9, in <module>
        load_entry_point('cookiecutter==0.9.0', 'console_scripts', 'cookiecutter')()
      File "/opt/anaconda/lib/python2.7/site-packages/click-3.3-py2.7.egg/click/core.py", line 610, in __call__
        return self.main(*args, **kwargs)
      File "/opt/anaconda/lib/python2.7/site-packages/click-3.3-py2.7.egg/click/core.py", line 590, in main
        rv = self.invoke(ctx)
      File "/opt/anaconda/lib/python2.7/site-packages/click-3.3-py2.7.egg/click/core.py", line 782, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/opt/anaconda/lib/python2.7/site-packages/click-3.3-py2.7.egg/click/core.py", line 416, in invoke
        return callback(*args, **kwargs)
      File "/Users/alexp/Projects/cookiecutter/cookiecutter/cli.py", line 70, in main
        cookiecutter(template, checkout, no_input)
      File "/Users/alexp/Projects/cookiecutter/cookiecutter/main.py", line 95, in cookiecutter
        extra_context=extra_context,
      File "/Users/alexp/Projects/cookiecutter/cookiecutter/generate.py", line 43, in generate_context
        file_handle = open(context_file)
    IOError: [Errno 2] No such file or directory: u'cookiecutter-pypackage/cookiecutter.json'

This error shows up if I either do pip install or with the git repo locally.  Naturally, it makes a bit of sense.  There is no directory ~/Projects/cookiecutter-pypackage/.  

However, and perhaps I'm making a poor assumption about usage, it seems to me if I clone or otherwise create a cookiecutter and it's sitting in cookiecutters_dir, it'd be nice to just refer to it as I did above.  For my usage, if I create a cookiecutter, I don't particularly want it sitting around a Projects directory, especially if I have multiple project directories for personal and organizational purposes.

In order to do this, I added three lines to main.py in my fork (see below) and it seems to work.  I did it as an `elif` to try to preserve the possibility of a lack of a cookiecutters_dir.  I have not written a test for this and admittedly I don't really know how.  I will likely just use my fork with this modificationgoing forward but I wanted to let the developer crew know about this.  

Cheers.

    # cookiecutter.main
    ...
    # TODO: find a better way to tell if it's a repo URL
    if 'git@' in template or 'https://' in template:
        repo_dir = clone(
            repo_url=template,
            checkout=checkout,
            clone_to_dir=config_dict['cookiecutters_dir'],
            no_input=no_input
        )
    #### Added these three lines
    elif 'cookiecutters_dir' in config_dict:
        cookiecutters_dir = os.path.expanduser(config_dict['cookiecutters_dir'])
        repo_dir = os.path.join(cookiecutters_dir,template)
    else:
        # If it's a local repo, no need to clone or copy to your
        # cookiecutters_dir
        repo_dir = template

### PR DESCRIPTION
**Pull request by michaeljoseph**
Existing cookiecutters in `~/.cookiecutters` can now be referenced by directory name.

Fixes #385.

### CODE REVIEW

File: cookiecutter/repository.py [hunk 1]
```diff
@@ -76,22 +76,27 @@ def determine_repo_dir(template, abbreviations, clone_to_dir, checkout,
...
             clone_to_dir=clone_to_dir,
             no_input=no_input,
         )
-    else:
-        # If it's a local repo, no need to clone or copy to your
-        # cookiecutters_dir
-        repo_dir = template
+        repository_candidates.append(cloned_repo)
+
+    repository_candidates.extend([
```
Comment (hackebrot) [inline cookiecutter/repository.py]: I think it would be better to have this only when ``is_repo_url`` evaluates to false.

File: cookiecutter/repository.py [hunk 2]
```diff
@@ -76,22 +76,27 @@ def determine_repo_dir(template, abbreviations, clone_to_dir, checkout,
...
-    if repository_has_cookiecutter_json(repo_dir):
-        return repo_dir
+    for repo_candidate in repository_candidates:
+        if repository_has_cookiecutter_json(repo_candidate):
+            return repo_candidate
 
     raise RepositoryNotFound(
         'The repository {} could not be located or does not contain '
-        'a "cookiecutter.json" file.'.format(repo_dir)
+        'a "cookiecutter.json" file.'.format(template)
```
Comment (hackebrot) [inline cookiecutter/repository.py]: It would be cool to extend this error message and list all the directories that have been checked, which would be the items in ``repository_candidates``.

File: tests/repository/test_determine_repo_dir_clones_repo.py [hunk 3]
```diff
@@ -57,12 +58,12 @@ def test_repository_url_with_no_context_file(mocker, template_url):
         repository.determine_repo_dir(
             template_url,
             abbreviations={},
-            clone_to_dir=None,
+            clone_to_dir=user_config_data['cookiecutters_dir'],
```
Comment (hackebrot) [inline tests/repository/test_determine_repo_dir_clones_repo.py]: Hmm why is this necessary?

File: tests/repository/test_determine_repo_dir_finds_existing_cookiecutter.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import os
+
+from cookiecutter import repository
+
+
+def test_should_find_existing_cookiecutter(user_config_data):
+    template = 'cookiecutter-pytest-plugin'
+    cookiecutters_dir = user_config_data['cookiecutters_dir']
+    cloned_template_path = os.path.join(cookiecutters_dir, template)
+    os.mkdir(cloned_template_path)
```
Comment (hackebrot) [inline tests/repository/test_determine_repo_dir_finds_existing_cookiecutter.py]: Let's move this to setup code in a fixture.

File: cookiecutter/repository.py (hunk 1 above)
Comment (michaeljoseph) [inline cookiecutter/repository.py]: Agreed.

File: cookiecutter/repository.py (hunk 2 above)
Comment (michaeljoseph) [inline cookiecutter/repository.py]: Done, not that happy with the actual message, so improvements are welcome.

File: tests/repository/test_determine_repo_dir_clones_repo.py (hunk 3 above)
Comment (michaeljoseph) [inline tests/repository/test_determine_repo_dir_clones_repo.py]: I switched this back now that we're not always checking if there's an existing clone.

File: tests/repository/test_determine_repo_dir_finds_existing_cookiecutter.py (hunk 4 above)
Comment (michaeljoseph) [inline tests/repository/test_determine_repo_dir_finds_existing_cookiecutter.py]: :+1:

**Comment by hackebrot** [comment]
Hey @michaeljoseph!

Happy to merge this. Must have missed the notification on new commits again 😞 

Any objections?

**Comment by pydanny** [comment]
Merge!!!

**Comment by hackebrot** [comment]
@pydanny 😀

**Comment by hackebrot** [comment]
Great work @michaeljoseph! 🍪

**Comment by michaeljoseph** [comment]
Thanks guys :smile_cat:
